### ISSUE TITLE
Repro Structure

### ISSUE THREAD

**Issue by pankajkr23** [The-School-of-AI/LLM#383]
# How Training & Coresets Pipelines Upload Artifacts to the Experiment Registry (Step-by-Step)

This document explains **how artifacts from the coresets pipeline and training pipeline are written into the canonical S3 Experiment Registry**, what **code Team 19 must build**, and **how other teams will use it**.

This is intended to be pasted directly into a GitHub Issue.

---

## Step 0 — Core Principle (Do Not Skip)

**Teams must never write directly to S3. ****

All provenance, metadata capture, directory layout, and uploads are handled by a **shared reproducibility SDK** owned by **Team 19**.

Other teams:
- import the SDK
- call a small set of APIs
- focus only on their domain logic (training / coresets)

This guarantees:
- correct structure
- immutability
- complete provenance
- zero ad-hoc S3 code

---

## Step 1 — Shared Repository Owned by Team 19

Create a single shared repo (example name):

p19/feat/repro-dev

This repo is:
- versioned
- reviewed
- imported by all pipelines (training, coresets, evaluation)

Teams **must not copy or fork** this logic.

---

## Step 2 — Code Structure to Be Implemented (Team 19)

repro-registry/
├── repro/
│   ├── init.py
│   ├── ids.py              # run_id / coreset_run_id generation
│   ├── git.py              # commit hash, repo cleanliness checks
│   ├── config.py           # config freezing + hashing
│   ├── env.py              # environment fingerprint capture
│   ├── seeds.py            # RNG seed capture
│   ├── manifest.py         # Pydantic schemas for manifests
│   ├── s3.py               # controlled S3 writer (no overwrite/delete)
│   ├── checksums.py        # artifact integrity
│   └── registry.py         # orchestration (start/finalize runs)
│
├── repro_cli/
│   └── main.py             # optional CLI helpers
│
├── examples/
│   ├── training_example.py
│   └── coreset_example.py
│
├── pyproject. toml
└── README.md

This is the **only interface** other teams should depend on.

---

## Step 3 — Two Supported Pipelines (No Others)

There are exactly **two canonical pipelines** supported by the registry:

1. **Coreset pipeline**
2. **Training pipeline**

They are parallel but distinct.

---

### Goal
Run `coresets_builder.py`, produce shards and metadata, and upload to:

s3:///coresets/<coreset_run_id>/

### How the team uses the SDK

```python
from repro.registry import start_coreset_run, finalize_run
from repro.seeds import set_all_seeds

def main(config):
    ctx = start_coreset_run(
        config=config,
        pipeline="coresets"
    )

    set_all_seeds(config.seed)

    shards = build_coresets(config)
    stats = compute_stats(shards)

    ctx.save_output("outputs/shards/", shards)
    ctx.save_output("outputs/metadata/stats.json", stats)

    finalize_run(ctx)
````

### What the SDK does automatically

- Generates coreset_run_id
- Captures git commit, frozen config, config hash
- Captures environment fingerprint and seeds
- Creates a local run directory:

.repro_tmp/coresets/<coreset_run_id>/

- Writes manifest.json
- Uploads all artifacts to S3
- Uploads the manifest last (atomic completion)

The team never interacts with S3 directly.

### Goal
Run PyTorch / Ray training, produce logs and checkpoints, and upload to:

s3://<registry-bucket>/runs/<run_id>/

### How the team uses the SDK
```python
from repro.registry import start_training_run, finalize_run
from repro.seeds import set_all_seeds

def main(config):
    ctx = start_training_run(
        config=config,
        pipeline="training",
        coreset_run_id=config.coreset_run_id
    )

    set_all_seeds(config.seed)

    model = build_model(config)

    for step in range(config.steps):
        loss = train_step(model)
        ctx.log_metric("loss", loss, step)

        if step % config.ckpt_every == 0:
            ctx.save_checkpoint(model, step)

    finalize_run(ctx)
````

### What the SDK does automatically

- Generates run_id
- Records coreset_run_id
- Captures git commit, frozen config, config hash
- Captures environment fingerprint and seeds
- Creates:

.repro_tmp/runs/<run_id>/

- Writes logs, checkpoints, and metrics deterministically
- Uploads artifacts immutably to S3

## Step 6 — Context Object (ctx) Responsibilities

The ctx object is the only write interface exposed to teams.

### Supported operations:

- ctx.save_output(path, obj)
- ctx.save_checkpoint(model, step)
- ctx.log_metric(name, value, step)

### Internally, it:

- writes only to the local run directory
- enforces deterministic paths
- forbids overwrites
- prepares data for immutable upload

⸻

### The SDK enforces:

- No overwrites (PutObject only on new keys)
- No deletes
- One prefix per run
- Manifest uploaded last

This matches IAM and immutability guarantees.

⸻

## Step 8 — GitHub Actions Integration

Both training and coresets pipelines run via GitHub Actions.

Example:
jobs:
  run:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - uses: actions/setup-python@v5

      - name: Install dependencies
        run: pip install repro-registry torch ray

      - name: Run pipeline
        env:
          AWS_ROLE_ARN: registry-writer
        run: |
          python train.py --config configs/train.yaml

### GitHub Actions:

- Assumes the IAM role
- The SDK handles all registry interactions
